firmware/console/binary/tunerstudio.cpp

```diff
@@ -751,85 +720,88 @@ int tunerStudioHandleCrcCommand(ts_channel_s *tsChannel, char *data, int incomin
-		uint16_t page = *(uint16_t *) data;
-		handleBurnCommand(tsChannel, TS_CRC, page);
-	} else if (command == TS_READ_COMMAND) {
-		uint16_t page = *(uint16_t *) data;
-		uint16_t offset = *(uint16_t *) (data + 2);
-		uint16_t count = *(uint16_t *) (data + 4);
-		handlePageReadCommand(tsChannel, TS_CRC, page, offset, count);
-	} else if (command == 't' || command == 'T') {
-		handleTestCommand(tsChannel);
-	} else if (command == TS_IO_TEST_COMMAND) {
-		uint16_t subsystem = SWAP_UINT16(*(uint16_t*)&data[0]);
-		uint16_t index = SWAP_UINT16(*(uint16_t*)&data[2]);
-
-		if (engineConfiguration->debugMode == DBG_BENCH_TEST) {
-			tsOutputChannels.debugIntField1++;
-			tsOutputChannels.debugIntField2 = subsystem;
-			tsOutputChannels.debugIntField3 = index;
-
+		break;
+	case TS_PAGE_COMMAND:
+		handlePageSelectCommand(tsChannel, TS_CRC, data16[0]);
+		break;
+	case TS_GET_STRUCT:
+		handleGetStructContent(tsChannel, data16[0], data16[1]);
+		break;
+	case TS_GET_FILE_RANGE:
+		handleReadFileContent(tsChannel, data16[0], data16[1], data16[2]);
+		break;
+	case TS_CHUNK_WRITE_COMMAND:
+		currentPageId = data16[0];
+		handleWriteChunkCommand(tsChannel, TS_CRC, data16[1], data16[2], data + sizeof(TunerStudioWriteChunkRequest));
+		break;
+	case TS_SINGLE_WRITE_COMMAND:
+		{
+			uint8_t value = data[4];
+			handleWriteValueCommand(tsChannel, TS_CRC, data16[0], data16[1], value);
 		}
+		break;
+	case TS_CRC_CHECK_COMMAND:
+		handleCrc32Check(tsChannel, TS_CRC, data16[0], data16[1], data16[2]);
+		break;
+	case TS_BURN_COMMAND:
+		handleBurnCommand(tsChannel, TS_CRC, data16[0]);
+		break;
+	case TS_READ_COMMAND:
+		handlePageReadCommand(tsChannel, TS_CRC, data16[0], data16[1], data16[2]);
+		break;
+	case TS_TEST_COMMAND:
+		[[fallthrough]];
+	case 'T':
+		handleTestCommand(tsChannel);
+		break;
+	case TS_IO_TEST_COMMAND:
+		{
+			uint16_t subsystem = SWAP_UINT16(data16[0]);
+			uint16_t index = SWAP_UINT16(data16[1]);
+
+			if (engineConfiguration->debugMode == DBG_BENCH_TEST) {
+				tsOutputChannels.debugIntField1++;
+				tsOutputChannels.debugIntField2 = subsystem;
+				tsOutputChannels.debugIntField3 = index;
+			}
 
 #if EFI_PROD_CODE
-		executeTSCommand(subsystem, index);
+			executeTSCommand(subsystem, index);
 #endif /* EFI_PROD_CODE */
-		sendOkResponse(tsChannel, TS_CRC);
-	} else {
+			sendOkResponse(tsChannel, TS_CRC);
+		}
+		break;
+	default:
 		tunerStudioError("ERROR: ignoring unexpected command");
 		return false;
 	}
+
 	return true;
 }
 
 void startTunerStudioConnectivity(void) {
-	if (sizeof(persistent_config_s) != getTunerStudioPageSize(0))
-		firmwareError(CUSTOM_OBD_TS_PAGE_MISMATCH, "TS page size mismatch: %d/%d", sizeof(persistent_config_s), getTunerStudioPageSize(0));
-
-	if (sizeof(TunerStudioOutputChannels) != TS_OUTPUT_SIZE)
-		firmwareError(CUSTOM_OBD_TS_OUTPUT_MISMATCH, "TS outputs size mismatch: %d/%d", sizeof(TunerStudioOutputChannels), TS_OUTPUT_SIZE);
+	// Assert tune & output channel struct sizes
```

Comment:
The result of these comparisons is known at compile time (ie, it won't even compile a branch here) - so we can throw the error at compile time. (#861)